content/lectures/lec03.md

```diff
@@ -7,4 +7,192 @@ slides: ""
+
+```
+$ docker run -it \
+  --name=letsencrypt \
+  --cap-add=NET_ADMIN \
+  -e PUID=1000 \
+  -e PGID=1000 -e TZ=Europe/London \
+  -e URL=walthome.duckdns.org \
+  -e SUBDOMAINS=wildcard \
+  -e VALIDATION=duckdns \
+  -e DUCKDNSTOKEN=youwillnevergetthis \
+  -e EMAIL=pwpon500@gmail.com \
+  -v ./test:/config \
+  --restart unless-stopped \
+  linuxserver/letsencrypt
+```
+
+Maybe this is easy for the computer to read, but for humans there are numerous drawbacks to writing long console commands like this. For one, they're inaccessible and confusing to people who have less experience with the Docker CLI. Second, how are we going to keep track of this? Perhaps we could throw it in a shell script and commit it to our repository, but shell scripts can be cumbersome to edit and do not always interface well with other programming languages.
+
+## Run System Wishlist
+
+So, if we're unhappy with the Docker CLI as our interface for running containers, let's think about what we want out of a run system:
+
+- Reproducible (should be easy to share)
+- Easy to understand what's going on (human readable)
+- Confidence in making changes to config
+- Resource lifecycle management (we can set parameters that manage our containers)
+
+If we look at the example from the previous section, we can see that apart from resource lifecycle management, we have nothing else from our wishlist.
+
+## Docker Compose
+
+The improvement here is Docker Compose. This is a tool that allows us to "compose" multiple Docker containers together, and manage their lifecycles. Docker Compose manages the lifecycle of multiple containers at the same time, handles all runtime settings for us (environment variables, port maps, volumes, etc), and can be started, stopped and inspected with simple shell commands. Now, if we look back at our wishlist, we meet all of the criteria.
+
+## YAML (YAML Ain't Markup Language)
+
+What does Docker compose actually look like? The configuration is written in a language called YAML. This is great for a lot of reasons: YAML is a simple and human-readable config language, it is standard across many infrastructure tools like Kubernetes and Continuous Integration, and it has types which will prevent us from making some major mistakes. The grammar of YAML is made up of these base types:
+
+- Scalar
+  - String
+  - Number
+  - Boolean
+- Dictionary 
+- List
+
+The top-level structure is always a dictionary, a list of key/value pairs. If you're familiar with JSON, then this should seem familiar. While there are only a few types, by composing these data types we can be quite expressive.
+
+### Scalars
+
+#### Strings
+
+For strings we don't need quotes, however it is generally considered best practice to quote your strings. If you include numbers in your strings and don't quote them, the compiler interprets it as a number, so it's always safer to add quotes.
+```yaml
+armaan: "Tobaccowalla"
+peyton: Walters
+```
+
+#### Multi-line Strings 
```

Comment:
Wait, apparently I screwed up the terminology a bit here. `>` denotes a folded style (where newlines are replaced with spaces) and `|` represents the literal style. Chomp is when you include a `-` after the pipe or bracket to remove the last newline from the result. [Helpful link](https://yaml-multiline.info/) we could use

Reply:
Isn't that what we said?